schedulerlatency: implement linear interpolation for percentiles

aadityasondhi opened the pull request:
This patch improves the percentile calculation by using linear interpolation between bucket boundaries instead of using the mid-point. The patch also adds tests to compare values of the scheduler histograms to those found in util/metric.go which are based on the prometheus implementation to maintain consistency among metrics.

Fixes #89829

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94193)
<!-- Reviewable:end -->

aadityasondhi commented on pkg/util/schedulerlatency/sampler.go:
```diff
@@ -332,15 +333,10 @@ func percentile(h *metrics.Float64Histogram, p float64) float64 {
-	// computed off of go.scheduler_latency vs.
-	// admission.scheduler_latency_listener.p99_nanos.
+	// Find the target rank within bucket boundaries.
+	// NB: The 0.5 is added for rounding purposes; it helps in cases where
```
@irfansharif After doing this here, I am wondering whether my decision in util/metric.go was correct and if this is the strategy we want. By rounding it to the nearest int, I think we are assuming the data within the bucket is equally spaced. For example, if we have 5 samples in a bucket with a lower bound of 0 and an upper bound of 10. We assume the data is [0, 2.5, 5, 7.5, 10]. Here, 99th and 100th percentile would return 10 instead of 9.9 for the 99th. I initially thought this was ok, but not so sure anymore. It probably doesn't matter when buckets have large counts as this smooths itself out, but something to consider and we should keep both parts consistent with each other.

FWIW prometheus client does not do this: https://github.com/prometheus/prometheus/blob/d9162189/promql/quantile.go#L100

aadityasondhi committed:
schedulerlatency: implement linear interpolation for percentiles

This patch improves the percentile calculation by using linear
interpolation between bucket boundaries instead of using the mid-point.
The patch also adds tests to compare values of the scheduler histograms
to those found in util/metric.go which are based on the prometheus
implementation to maintain consistency among metrics.

Fixes #89829.

Release note: None

aadityasondhi reviewed (commented):
I did some cleanup based on your suggestions. PTAL, and let me know if there are other things I should fix.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94193#-:-NLgqazd0irf9a7I-0Km:bajzplg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[`pkg/util/schedulerlatency/sampler.go` line 337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94193#-NJw1DZUQGmtIZBIy8v2-r1-337:-NLgqDRDDxaxmx6UpENb:b-18any8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a9896ba6ccd5dc4cdf3f96c9a025ff16741ebb1/pkg/util/schedulerlatency/sampler.go#L337)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

We shouldn't need this rounding (didn't understand it even back in https://github.com/cockroachdb/cockroach/pull/87883#pullrequestreview-1111140344). Also, the uint64 type cast is having you lose precision. Pulled this branch down and tried the following, which worked better with the existing unit tests:

```
rank := float64(total)*p - float64(total-cumulative)
interpolator := rank / float64(h.Counts[i])
return start + (end-start)*interpolator
```

Is `rank` the best/most precise name here? Given the return values are floating points, use `require.InDelta` instead of `require.Equal` in the supporting test. Could you also improve the commentary for this math? I found it difficult to follow. It looks correct to me after the following (verbose, sorry!) derivation:

```
	// Want the `p * total`-th value, using something like:
	//    bucket-start + (bucket-end - bucket-start) * linear_interoplator
	//
	// For this next example:
	//  - Percentile of value at end-of-bucket: p100
	//  - Percentile of value at start-of-bucket: p90
	//  - Target percentile: p95
	//
	// In our example:
	//    linear interpolated p95 = bucket-start + (bucket-end - bucket-start) * 0.5
	//
	// Where 0.5 == (95 - 90) / (100 - 90) == 5 / 10 = 0.5. So:
	//    linear_interpolator = (p - pstart) / (pend - pstart)
	//
	// Taking each term:
	//
	// 1.
	//   (pend - pstart) == (total - cumulative + h.Counts[i] - total + cumulative) / total
	//   (pend - pstart) == h.Counts[i] / total
	//
	// 2.
	//   (p - pstart) == p - (total-cumulative) / total
	//   (p - pstart) == (total*p - (total-cumulative)) / total
	//
	// Simplifying 1 / 2:
	//
	//   (p - pstart) / (pend - pstart) == ((total*p - (total-cumulative)) / total ) / (h.Counts[i] / total)
	//   (p - pstart) / (pend - pstart) == ((total*p - (total-cumulative)) / total ) * (total / h.Counts[i])
	//   (p - pstart) / (pend - pstart) == ((total*p - (total-cumulative)) ) * (1 / h.Counts[i])
	//   (p - pstart) / (pend - pstart) == (total*p - (total-cumulative)) / h.Counts[i]
```

I'm likely missing some more obvious way you got to this form, but could you add it as a comment?
</blockquote></details>

Removed casting and rounding. Also added an explanation of how I derived the formula. It is less mathematical than what you did, but if you prefer your version, I am happy to include that instead.
___
*[`pkg/util/schedulerlatency/scheduler_latency_test.go` line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94193#-NJw1DZUQGmtIZBIy8v3-r2-124:-NLgqY9T0NTOKok3xHim:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1683b29394bb0f885da38d4f0617331252cda4b/pkg/util/schedulerlatency/scheduler_latency_test.go#L124)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

See comment above. After dropping the 0.5, use `require.InDelta` instead. After the uint casting change, some of these values will change. 
</blockquote></details>

Done.
___
*[`pkg/util/schedulerlatency/scheduler_latency_test.go` line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94193#-NJw1DZUQGmtIZBIy8v3-r2-130:-NLgqZPl3QxSUSzSbFoq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1683b29394bb0f885da38d4f0617331252cda4b/pkg/util/schedulerlatency/scheduler_latency_test.go#L130)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Good idea to compare this to the prometheus impl. Could you take this out into a separate unit test? Either in this file+package, or in the prometheus one.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> With the comment suggested below.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94193#-:-NLvP8hSCUBUQa9jRNow:b-yu9ysr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`pkg/util/schedulerlatency/sampler.go` line 337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94193#-NJw1DZUQGmtIZBIy8v2-r1-337:-NLvM5gg2j0y0dDq7Fqk:b-nadli3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a9896ba6ccd5dc4cdf3f96c9a025ff16741ebb1/pkg/util/schedulerlatency/sampler.go#L337)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Removed casting and rounding. Also added an explanation of how I derived the formula. It is less mathematical than what you did, but if you prefer your version, I am happy to include that instead.
</blockquote></details>

I think the commentary is still slightly ambiguous. We have this first formula: `rank = percentile * count`, which when "scoped" to just a bucket, gives us `subset_percentile = subset_rank / subset_count`. But then we're using `subset_percentile` as an interpolation term in `bucket_size * subset_percentile` (`bucket_size` here being different from `bucket_count`). Maybe I'm the only confused reader (something about "subset percentile" reads oddly), but I recommend this small amendment instead.

```
// Goal: We want to linear approximate the "p * total"-th value from the histogram.
//      - p * total is the ordinal rank (or simply, rank) of the target value
//        within the histogram bounds.
//
// Step 1: Find the bucket[i] which contains the target value.
//      - This will be the largest bucket[i] where the left-bound cumulative
//        count of bucket[i] <= p*total.
//
// Step 2: Determine the rank within the bucket (subset of histogram).
//      - Since the bucket is a subset of the original data set, we can simply
//        subtract left-bound cumulative to get the "subset_rank".
//
// Step 3: Use the "subset_rank" to interpolate the target value.
//      - So we want to interpolate the "subset_rank"-th value of the subset_count
//        that lies between [bucket_start, bucket_end). This is:
//        bucket_start + (bucket_start - bucket_end) * linear_interpolator
//        Where linear_interpolator = subset_rank / subset_count
```


<!-- Sent from Reviewable.io -->

irfansharif reviewed (approved):
Oops, the last rubber stamp didn't go through.

aadityasondhi commented:
bors r+

aadityasondhi commented:
bors r-

aadityasondhi commented:
bors r+
